### PR TITLE
Improved ScriptEngine debug messages

### PR DESCRIPTION
**Pull request by Atlante45**
- Check local scripts syntax before evaluating them.
- Display backtrace in case of exceptions.
- Make sure to pass a filename to `evaluate` for easier exception tracking.
- Cleaned up handler callbacks code.

Since we're checking the syntax before running a local script now, it might break scripts that had a syntax error in a very rarely used code path. I fixed 2 scripts and made sure all the script in defaultScript.js ran correctly.

But if you notice a script is not running, check your log for those tags: `[SyntaxError]` `[UncaughtException]`

New messages:
```
[10/26 11:43:06] [WARNING] [SyntaxError] Expected `identifier' in file:///Users/clement/hifi/examples/libraries/entityCameraTool.js:569(5)
```

```
[10/26 11:47:16] [WARNING] [UncaughtException] ReferenceError: Can't find variable: event in /Users/clement/hifi/examples/edit.js:1719
[10/26 11:47:16] [WARNING] [Backtrace]
[10/26 11:47:16] [WARNING]     <anonymous>(name = 'Show in Marketplace') at /Users/clement/hifi/examples/edit.js:1719
[10/26 11:47:16] [WARNING]     <global>() at /Users/clement/hifi/examples/edit.js:1847
```

### CODE REVIEW

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1e4a414f9df210ffc32b022c132cc49f39ac4127 available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 1]
```diff
@@ -662,6 +650,9 @@ void ScriptEngine::run() {
         }
         lastUpdate = now;
 
+        if (!checkExceptions(this, _fileNameString)) {
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: why did you move this below the emit update() call? That's what it looks like. Is there a reason you did that?

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 2]
```diff
@@ -898,10 +889,41 @@ void ScriptEngine::load(const QString& loadFile) {
...
+        const auto line = QString::number(syntaxCheck.errorLineNumber());
+        const auto column = QString::number(syntaxCheck.errorColumnNumber());
+        const auto message = QString("[SyntaxError] %1 in %2:%3(%4)").arg(error, program.fileName(), line, column);
+        qCWarning(scriptengine) << qPrintable(message);
+        return false;
+    }
+    return true;
+}
+
+bool ScriptEngine::checkExceptions(QScriptEngine* engine, const QString& fileName) {
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: why are you passing the QScriptEngine here? The QScriptEngine for any ScriptEngine should always be "this".. this really shouldn't be a static method.

File: libraries/script-engine/src/ScriptEngine.cpp (hunk 1 above)
Comment (Atlante45) [inline libraries/script-engine/src/ScriptEngine.cpp]: Update callbacks can throw too, so we want to check after it ran.
Actually we might want to check twice, after proccessEvent and after update, that's the only 2 failure point were we ran JS code in that loop AFAICT.

File: libraries/script-engine/src/ScriptEngine.cpp (hunk 2 above)
Comment (Atlante45) [inline libraries/script-engine/src/ScriptEngine.cpp]: That's to be able to run on the sandbox engine, even though I just realized I forgot to add that check.

**Comment by hifi-gustavo** [comment]
Build console output for commit: 1c048880ee2d3f33f38c20defc74dedf80f1cba5 available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: e73d50901903b33785c10dec19cbe487aac2c594 available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 380139364c4b2e0f74c6881f5cc809094d504c0a available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 3]
```diff
@@ -79,11 +79,41 @@ void avatarDataFromScriptValue(const QScriptValue &object, AvatarData* &out) {
 QScriptValue inputControllerToScriptValue(QScriptEngine *engine, AbstractInputController* const &in) {
     return engine->newQObject(in);
 }
-
 void inputControllerFromScriptValue(const QScriptValue &object, AbstractInputController* &out) {
     out = qobject_cast<AbstractInputController*>(object.toQObject());
 }
 
+bool checkSyntax(const QScriptProgram& program) {
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: should we make this "static" so that the symbols are not exposed outside of the scope of this file? not a big deal but for completeness... like you said, it's not needed outisde of here.

File: libraries/script-engine/src/ScriptEngine.cpp [hunk 4]
```diff
@@ -79,11 +79,41 @@ void avatarDataFromScriptValue(const QScriptValue &object, AvatarData* &out) {
...
+        const auto line = QString::number(syntaxCheck.errorLineNumber());
+        const auto column = QString::number(syntaxCheck.errorColumnNumber());
+        const auto message = QString("[SyntaxError] %1 in %2:%3(%4)").arg(error, program.fileName(), line, column);
+        qCWarning(scriptengine) << qPrintable(message);
+        return false;
+    }
+    return true;
+}
+
+bool checkExceptions(QScriptEngine& engine, const QString& fileName) {
```
Comment (ZappoMan) [inline libraries/script-engine/src/ScriptEngine.cpp]: same?

**Comment by hifi-gustavo** [comment]
Build console output for commit: ffe73348fb9157186005b9e93e8ef24cb96a5a99 available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

File: libraries/script-engine/src/ScriptEngine.cpp (hunk 3 above)
Comment (Atlante45) [inline libraries/script-engine/src/ScriptEngine.cpp]: Good point

**Comment by hifi-gustavo** [comment]
Build console output for commit: c0138bd189542116b323c1e75c75f68b7afea8c1 available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg

**Comment by hifi-gustavo** [comment]
Build console output for commit: 32f88f3fb6a664d5e664db51415fd8ecc0dd007d available at: https://gist.github.com/d57b4697db6bd2f7ae7c<br/><br/>Installers available for download at: <br/>Windows: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.exe<br/>OS X: http://s3.amazonaws.com/hifi-public/pr-builds/dev-interface-6182.dmg
